### PR TITLE
plugin/docker: add multi-stage target config

### PR DESCRIPTION
**Pull request by radriaanse**
Docker and Img can stop the build at the specified target stage.
If not set the full Dockerfile will be build.

Closes: #1969

### CODE REVIEW

**Comment by radriaanse** [comment]
Noticed that the CI was failing on website-mdx, but when I ran `make gen/website-mdx` as suggest I got a ton of changes with single->double quote and indenting.
Not sure if something on my end, but I've just skipped 'em and only included builder-docker.mdx.

**Comment by krantzinator** [comment]
@radriaanse can you add a changelog entry? You'll see lots of examples in our `.changelog` dir. I'd mark this one as an `improvement`. The title of the file should be `1992.txt` to match the PR number.

**Comment by briancain** [comment]
@radriaanse - ah yes, we should document how to set this up. Right now we have a node package that trims out all of the extra single quote/double quote changes from running that command from our website team. Kind of a weird situation, but just committing the docker plugin markdown looks fine to me! 👍🏻

File: builtin/docker/builder.go [hunk 1]
```diff
@@ -322,6 +328,7 @@ func (b *Builder) buildWithDocker(
 	tag string,
 	platform string,
 	buildArgs map[string]*string,
+	Target string,
```
Comment (briancain) [inline builtin/docker/builder.go:331]: This should be `target`

```suggestion
	target string,
```

File: builtin/docker/builder.go [hunk 2]
```diff
@@ -370,6 +377,7 @@ func (b *Builder) buildWithDocker(
 		Remove:     true,
 		Platform:   platform,
 		BuildArgs:  buildArgs,
+		Target:     Target,
```
Comment (briancain) [inline builtin/docker/builder.go:380]:
```suggestion
		Target:     target,
```

File: builtin/docker/img.go [hunk 3]
```diff
@@ -24,6 +24,7 @@ func (b *Builder) buildWithImg(
 	contextDir string,
 	tag string,
 	buildArgs map[string]*string,
+	Target string,
```
Comment (briancain) [inline builtin/docker/img.go:27]:
```suggestion
	target string,
```

File: builtin/docker/img.go [hunk 4]
```diff
@@ -38,6 +39,7 @@ func (b *Builder) buildWithImg(
 		"build",
 		"-f", dockerfilePath,
 		"-t", tag,
+		"--target", Target,
```
Comment (briancain) [inline builtin/docker/img.go:42]:
```suggestion
		"--target", target,
```

**Review by briancain — CHANGES_REQUESTED** [review]
Thanks for opening the PR to add this @radriaanse! I've left a few comments just to fix the variable casing for `target`. A question I had was what happens when we pass `--target ""` to docker or img? Does it act like a no-op and ignore the flag?

**Comment by radriaanse** [comment]
Wooeps, I've missed that; fixed! @briancain 

> A question I had was what happens when we pass --target "" to docker or img? Does it act like a no-op and ignore the flag?

Yes it does indeed just ignore it / builds all stages:

```Dockerfile
FROM docker.io/library/alpine as stage1
ENV a=1

FROM stage1 as stage2
ENV a=2
```

Without a target parameter:

```
waypoint [1969-docker-target●] ./waypoint build
✓ Initializing Docker client...
✓ Building image...
 │ Step 2/4 : ENV a=1
 │  ---> Using cache
 │  ---> 9c52a1f1e65c
 │ Step 3/4 : FROM stage1 as stage2
 │  ---> 9c52a1f1e65c
 │ Step 4/4 : ENV a=2
 │  ---> Using cache
 │  ---> 4499cfcd9e74
 │ Successfully built 4499cfcd9e74
 │ Successfully tagged waypoint.local/1992:latest
✓ Injecting Waypoint Entrypoint...
```

With a target parameter:
```
waypoint [1969-docker-target●] ./waypoint build
✓ Initializing Docker client...
✓ Building image...
 │ Step 1/2 : FROM docker.io/library/alpine:3.13.5 as stage1
 │  ---> 6dbb9cc54074
 │ Step 2/2 : ENV a=1
 │  ---> Using cache
 │  ---> 9c52a1f1e65c
 │ Successfully built 9c52a1f1e65c
 │ Successfully tagged waypoint.local/1992:latest
✓ Injecting Waypoint Entrypoint...
```

on the CLI
```
waypoint [1969-docker-target●] docker build -f test.Dockerfile --target "" .
Sending build context to Docker daemon  325.5MB
Step 1/4 : FROM docker.io/library/alpine:3.13.5 as stage1
 ---> 6dbb9cc54074
Step 2/4 : ENV a=1
 ---> Using cache
 ---> 9c52a1f1e65c
Step 3/4 : FROM stage1 as stage2
 ---> 9c52a1f1e65c
Step 4/4 : ENV a=2
 ---> Using cache
 ---> 4499cfcd9e74
Successfully built 4499cfcd9e74
```

```
waypoint [1969-docker-target●] img build -f test.Dockerfile -t docker.io/radriaanse/imgstage --target "" .
Building docker.io/radriaanse/imgstage:latest
Setting up the rootfs... this may take a bit.
WARN[0000] using host network as the default            
[+] Building 1.4s (2/3)                                                                                                                                                                                                                                                                                                                                                                              
[+] Building 1.5s (6/6) FINISHED                                                                                                                                                                                                                                                                                                                                                                     
 => [internal] load .dockerignore                                                                                                                                                                                                                                                                                                                                                               0.0s
 => => transferring context: 34B                                                                                                                                                                                                                                                                                                                                                                0.0s
 => [internal] load build definition from test.Dockerfile                                                                                                                                                                                                                                                                                                                                       0.1s
 => => transferring dockerfile: 36B                                                                                                                                                                                                                                                                                                                                                             0.0s
 => [internal] load metadata for docker.io/library/alpine:latest                                                                                                                                                                                                                                                                                                                                1.3s
 => CACHED [stage1 1/1] FROM docker.io/library/alpine@sha256:eb3e4e175ba6d212ba1d6e04fc0782916c08e1c9d7b45892e9796141b1d379ae                                                                                                                                                                                                                                                                   0.0s
 => => resolve docker.io/library/alpine@sha256:eb3e4e175ba6d212ba1d6e04fc0782916c08e1c9d7b45892e9796141b1d379ae                                                                                                                                                                                                                                                                                 0.0s
 => exporting to image                                                                                                                                                                                                                                                                                                                                                                          0.0s
 => => exporting layers                                                                                                                                                                                                                                                                                                                                                                         0.0s
 => => exporting manifest sha256:e0a74fbb475af1987799e25c915f1d7b450e609fd0009ccdb7fac18276ef5f2b                                                                                                                                                                                                                                                                                               0.0s
 => => exporting config sha256:341ca9a9cc40f27c74ba687660ce5512979e49c6c9b5e95327cae32a6f776c0d                                                                                                                                                                                                                                                                                                 0.0s
 => => naming to docker.io/radriaanse/imgstage:latest                                                                                                                                                                                                                                                                                                                                           0.0s
 => exporting cache                                                                                                                                                                                                                                                                                                                                                                             0.0s
 => => preparing build cache for export                                                                                                                                                                                                                                                                                                                                                         0.0s
Successfully built docker.io/radriaanse/imgstage:latest

waypoint [1969-docker-target●] img save docker.io/radriaanse/imgstage:latest | docker load
bc276c40b172: Loading layer [==================================================>]  2.813MB/2.813MB
Loaded image: radriaanse/imgstage:latest

waypoint [1969-docker-target●] docker run --rm docker.io/radriaanse/imgstage env | grep 'a='
a=2
```

**Pull request merged**